.github/workflows/src/verify-run-status.js

```diff
@@ -0,0 +1,174 @@
+import { extractInputs } from "./context.js";
+import { PER_PAGE_MAX } from "./github.js";
+
+/**
+ * @typedef {import('@octokit/plugin-rest-endpoint-methods').RestEndpointMethodTypes} RestEndpointMethodTypes
+ * @typedef {RestEndpointMethodTypes["checks"]["listForRef"]["response"]["data"]["check_runs"][number]} CheckRun
+ * @typedef {RestEndpointMethodTypes["actions"]["listWorkflowRunsForRepo"]["response"]["data"]["workflow_runs"][number]} WorkflowRun
+ */
+
+/**
+ * Given the name of a completed check run name and a completed workflow, verify 
+ * that both have the same conclusion. If conclusions are different, fail the 
+ * action.
+ * @param {import('github-script').AsyncFunctionArguments} AsyncFunctionArguments
+ */
+export async function verifyRunStatus({ github, context, core }) {
+  const checkRunName = process.env.CHECK_RUN_NAME;
+  if (!checkRunName) {
+    throw new Error("CHECK_RUN_NAME is not set");
+  }
+
+  const workflowName = process.env.WORKFLOW_NAME;
+  if (!workflowName) {
+    throw new Error("WORKFLOW_NAME is not set");
+  }
+
+  // Exit early when context is a check_run event and the check run does not
+  // match the checkRunName.
+  if (context.eventName == "check_run") {
+    const contextRunName = context.payload.check_run.name;
+    if (contextRunName !== checkRunName) {
+      core.notice(`Check run name (${contextRunName}) does not match input: ${checkRunName}`);
+      return;
+    }
+  }
+
+  const { head_sha } = await extractInputs(github, context, core);
+
+  const checkRun =
+    context.eventName == "check_run"
+      ? context.payload.check_run
+      : await getCheckRun(github, context, core, checkRunName, head_sha);
```

Comment:
To prevent errors, i think it's safer to explictly handle every eventName you want to support, and throw if you get an unexpected eventName.  In this case, I think you need to allow check_run, check_suite, and workflow_run.